specs/AppLifecycle/StateNotifications/AppLifecycle StateNotifications.md

```diff
@@ -0,0 +1,389 @@
+callback model for Cobalt. Additional WM-based behavior can be added in a later release, if customer
+demand indicates it would provide significant benefit.
+
+The net is that the Reunion PowerManager API will be a combination of WinRT PowerManager and Win32
+PowerSettingRegisterNotification. We will define new WinRT types (in
+Microsoft.ApplicationModel.PowerManager) that provide event handlers for apps to use.
+
+## Examples
+
+### Subscribing and responding to events
+
+The power notifications are relatively simple. The app provides a callback function for each
+notification it wants to receive, and when the event is raised the system invokes that callback. The
+app could start listening for the notifications at any time; it would most often do this early in
+its life, perhaps in WinMain - but the AppLifecycle mechanism does not impose any constraints on
+this, and the app is free to subscribe or unsubscribe at any time.
+
+```c++
+int APIENTRY wWinMain(
+    _In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance,
+    _In_ LPWSTR lpCmdLine, _In_ int nCmdShow)
+{
+    // Optionally, register callbacks for power/battery state changes.
+    PowerManager::BatteryStatusChanged([](auto&&...)
+        { OnBatteryStateChange(); });
+    PowerManager::PowerSupplyStatusChanged([](auto&&...)
+        { OnPowerStateChange(); });
+    PowerManager::EnergySaverStatusChanged([](auto&&...)
+        { OnEnergySaverStateChange(); });
+    PowerManager::RemainingChargePercentChanged([](auto&&...)
+        { OnRemainingChargePercentChange(); });
+    PowerManager::RemainingDischargeTimeChanged([](auto&&...)
+        { OnRemainingDischargeTimeChange(); });
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Standard Win32 window configuration/creation and message pump:
+    // ie, whatever the app would normally do - nothing new here.
+    LoadStringW(hInstance, IDS_APP_TITLE, szTitle, MAX_LOADSTRING);
+    LoadStringW(hInstance, IDC_CLASSIC, szWindowClass, MAX_LOADSTRING);
+    MyRegisterClass(hInstance);
+    if (!InitInstance(hInstance, nCmdShow))
+    {
+        return FALSE;
+    }
+    MSG msg;
+    while (GetMessage(&msg, nullptr, 0, 0))
+    {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    }
+    return (int)msg.wParam;
+}
+```
+
+In the app-defined callback, the app can do whatever it likes. Typically, it would fetch the current
+status of the particular feature (that is, the battery status, power supply status, energy saver
+status, and so on), and use that information to form its decision. For example, if battery is at a
+low power state and discharging, the app might choose to defer any non-critical background work, in
+the interests of being a good power/battery citizen.
+
+```c++
+void OnBatteryStateChange()
+{
+    BatteryStatus batteryState = PowerManager::BatteryStatus();
+    int remainingCharge = PowerManager::RemainingChargePercent();
```

Comment:
What does this return if there is no battery?

Reply:
RECOMMEND: Add documentation into the IDL/docs. The `state` has an enum that says `NotPresent`.  The remaining charge percent should probably be "all" when "not present" (desktop-class machine.)

SAMPLE: This sample does not run power-intensive work when on AC, for example.

Reply:
Fixed.